fixes #79 allowed key as null in converted message

shubhamvashisht opened the pull request.

ppatierno commented:
May you write a test for this one please ?

shubhamvashisht commented:
@ppatierno sure.

shubhamvashisht commented:
@ppatierno hey, I am confused what new test to add for this change. Any suggestions?

shubhamvashisht commented:
@ppatierno Added a test as requested.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java:
```diff
@@ -843,4 +845,13 @@ public void noPartitionsAvailable(TestContext context) {
+		AmqpDefaultMessageConverter messageConverter = new AmqpDefaultMessageConverter();
+		Message message = messageConverter.toMessage("0",record);
+		context.assertNull(message.getMessageAnnotations().getValue().get(Symbol.valueOf(AmqpBridge.AMQP_KEY_ANNOTATION)));
+	}
```
Thanks for the added tests ! A couple of comments here.
Out of curiosity ... why that kind of payload ? :-)
Then you should need in the same test method, the check for `null` key even for the other two converters (raw and json).

shubhamvashisht replied:
I have added test for jsonMessageConverter too and also corrected the payload. :D 
but the rawMessageConverter throws org.apache.qpid.proton.codec.DecodeException: Unknown constructor for same payload. i couldn't find the reason.

shubhamvashisht commented:
I have added test for jsonMessageConverter too and also corrected the payload.(works for json and default converter) :D
but the rawMessageConverter throws org.apache.qpid.proton.codec.DecodeException: Unknown constructor for same payload. i couldn't find the reason.

ppatierno commented:
@shubhamvashisht can you put the entire stack trace here ?

shubhamvashisht commented:
yes here. 
org.apache.qpid.proton.codec.DecodeException: Unknown constructor

	at org.apache.qpid.proton.codec.DecoderImpl.readObject(DecoderImpl.java:939)
	at org.apache.qpid.proton.message.impl.MessageImpl.decode(MessageImpl.java:593)
	at org.apache.qpid.proton.message.impl.MessageImpl.decode(MessageImpl.java:572)
	at io.strimzi.kafka.bridge.amqp.converter.AmqpRawMessageConverter.toMessage(AmqpRawMessageConverter.java:83)
	at io.strimzi.kafka.bridge.amqp.AmqpBridgeTest.ConvertedMessageWithNullKey(AmqpBridgeTest.java:861)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.vertx.ext.unit.junit.VertxUnitRunner.invokeTestMethod(VertxUnitRunner.java:95)
	at io.vertx.ext.unit.junit.VertxUnitRunner.lambda$invokeExplosively$0(VertxUnitRunner.java:114)
	at io.vertx.ext.unit.impl.TestContextImpl$Step.run(TestContextImpl.java:128)
	at io.vertx.ext.unit.impl.TestContextImpl$Step.access$200(TestContextImpl.java:30)
	at io.vertx.ext.unit.impl.TestContextImpl.run(TestContextImpl.java:237)
	at io.vertx.ext.unit.junit.VertxUnitRunner.invokeExplosively(VertxUnitRunner.java:130)
	at io.vertx.ext.unit.junit.VertxUnitRunner.access$000(VertxUnitRunner.java:39)
	at io.vertx.ext.unit.junit.VertxUnitRunner$1.evaluate(VertxUnitRunner.java:84)
	at io.vertx.ext.unit.junit.VertxUnitRunner$2.evaluate(VertxUnitRunner.java:196)
	at io.vertx.ext.unit.junit.VertxUnitRunner$3.evaluate(VertxUnitRunner.java:211)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at io.vertx.ext.unit.junit.VertxUnitRunner$2.evaluate(VertxUnitRunner.java:196)
	at io.vertx.ext.unit.junit.VertxUnitRunner$3.evaluate(VertxUnitRunner.java:211)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

ppatierno commented:
I think that something changed in the Qpid Proton and I have to figure out what.

Btw instead of having one test with same code more times, I think you should factor out a private method doing the common stuff and then create three test methods for the three converters. Mark with `@Ignore` the test related to the raw converter.

shubhamvashisht commented:
@ppatierno You may review the requested changes.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java:
```diff
@@ -843,4 +848,32 @@ public void noPartitionsAvailable(TestContext context) {
+	@Test
+	public void defaultMessageConverterNullKeyTest(TestContext context) {
+		AmqpDefaultMessageConverter defaultMessageConverter = new AmqpDefaultMessageConverter();
+		context.assertNull(convertedMessageWithNullKey(defaultMessageConverter));
```
Use `MessageConverter` for declaration.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java:
```diff
@@ -843,4 +848,32 @@ public void noPartitionsAvailable(TestContext context) {
+
+	@Test
+	public void jsonMessageConverterNullKeyTest(TestContext context) {
+		AmqpJsonMessageConverter jsonMessageConverter = new AmqpJsonMessageConverter();
```
Ditto as above.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java:
```diff
@@ -843,4 +848,32 @@ public void noPartitionsAvailable(TestContext context) {
+	@Ignore
+	@Test
+	public void rawMessageConverterNullKeyTest(TestContext context) {
+		AmqpRawMessageConverter rawMessageConverter = new AmqpRawMessageConverter();
```
Ditto as above.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/amqp/AmqpBridgeTest.java:
```diff
@@ -843,4 +848,32 @@ public void noPartitionsAvailable(TestContext context) {
+
+		//Record with a null key
+		ConsumerRecord<String,byte[]> record = new ConsumerRecord<String,byte[]>("mytopic",0,0,null,payload.getBytes());
+		Message message = (Message) messageConverter.toMessage("0",record);
```
Use space after every comma when specify parameters.

shubhamvashisht commented:
Changes are done. Thanks for reviewing.
